datafusion/src/physical_optimizer/repartition.rs

```diff
@@ -219,20 +384,64 @@ mod tests {
             "ParquetExec: limit=None, partitions=[x]",
         ];
 
-        assert_eq!(&trim_plan_display(&plan), &expected);
+        assert_optimized!(expected, plan);
         Ok(())
     }
 
     #[test]
-    fn repartition_ignores_limit() -> Result<()> {
-        let optimizer = Repartition {};
+    fn repartition_unsorted_limit() -> Result<()> {
+        let plan = limit_exec(filter_exec(parquet_exec()));
+
+        let expected = &[
+            "GlobalLimitExec: limit=100",
+            "LocalLimitExec: limit=100",
+            "FilterExec: c1@0",
+            // nothing sorts the data, so the local limit doesn't require sorted data either
+            "RepartitionExec: partitioning=RoundRobinBatch(10)",
+            "ParquetExec: limit=None, partitions=[x]",
+        ];
+
+        assert_optimized!(expected, plan);
+        Ok(())
+    }
+
+    #[test]
+    fn repartition_sorted_limit() -> Result<()> {
+        let plan = limit_exec(sort_exec(parquet_exec()));
+
+        let expected = &[
+            "GlobalLimitExec: limit=100",
+            "LocalLimitExec: limit=100",
+            // data is sorted so can't repartition here
+            "SortExec: [c1@0 ASC]",
+            "ParquetExec: limit=None, partitions=[x]",
+        ];
+
+        assert_optimized!(expected, plan);
+        Ok(())
+    }
+
+    #[test]
+    fn repartition_sorted_limit_with_filter() -> Result<()> {
+        let plan = limit_exec(filter_exec(sort_exec(parquet_exec())));
 
-        let optimized = optimizer.optimize(
-            hash_aggregate(limit_exec(filter_exec(limit_exec(parquet_exec())))),
-            &ExecutionConfig::new().with_target_partitions(10),
-        )?;
+        let expected = &[
+            "GlobalLimitExec: limit=100",
+            "LocalLimitExec: limit=100",
+            "FilterExec: c1@0",
+            // data is sorted so can't repartition here even though
```

Comment:
However, once you put a sort here then repartitioning can't happen without potentially getting wrong results